5.drug-dependency/scripts/4.drug_latent_dim_correlation.py

```diff
@@ -0,0 +1,263 @@
+model_save_dir = pathlib.Path("../4.gene_expression_signatures/saved_models")
+output_dir = pathlib.Path("results")
+output_dir.mkdir(parents=True, exist_ok=True)
+
+# Latent dimensions and model names to iterate over
+latent_dims = [2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 14, 16, 18, 20, 25, 30, 35, 40, 45, 50, 60, 70, 80, 90, 100, 150, 200]
+model_names = ["pca", "ica", "nmf", "vanillavae", "betavae", "betatcvae"]
+
+# File to store the combined correlation results
+final_output_file = output_dir / "combined_latent_drug_correlations.parquet"
+try:
+    combined_results_df = pd.read_parquet(final_output_file)
+    print(f"Loaded existing results from {final_output_file}")
+except FileNotFoundError:
+    # If the file doesn't exist, initialize an empty DataFrame
+    combined_results_df = pd.DataFrame()
+    print(f"No existing file found. Initialized empty DataFrame.")
+
+for num_components in latent_dims:
+    for model_name in model_names:
+        # Check if this model and latent dimension have already been processed
+        if not combined_results_df.empty:
+            if ((combined_results_df['model'] == model_name) & 
+                (combined_results_df['full_model_z'] == num_components)).any():
+                print(f"Skipping {model_name} with {num_components} dimensions as it is already processed.")
+                continue  # Skip to the next iteration if this combination is already present
+        
+        # Load the saved model
+        model_filename = model_save_dir / f"{model_name}_{num_components}_components_model.joblib"
+        if model_filename.exists():
+            print(f"Loading model from {model_filename}")
+            model = joblib.load(model_filename)
+            
+            if model_name in ["pca", "ica", "nmf"]:
+                # Extract the latent dimensions for these models
+                latent_df = sklearn_extract_latent_dimensions(model, model_name, dependency_df)
+            elif model_name == "betavae":
+                latent_df = extract_latent_dimensions(model, train_and_test_subbed_loader, metadata)
+            elif model_name == "betatcvae":
+                latent_df = tc_extract_latent_dimensions(model, train_and_test_subbed_loader, metadata)
+            elif model_name == "vanillavae":
+                latent_df = vvae_extract_latent_dimensions(model, train_and_test_subbed_loader, metadata)
+
+            latent_df.columns = ['ModelID'] + [f'z_{col}' if isinstance(col, int) else col for col in latent_df.columns[1:]]
+            # Perform Pearson correlation between latent dimensions and drug data
+            correlation_results_df = perform_correlation(latent_df, prism_df, model_name, num_components)
+            
+            # Concatenate results to the combined dataframe
+            combined_results_df = pd.concat([combined_results_df, correlation_results_df], ignore_index=True)
+        else:
+            print(f"Model file {model_filename} not found. Skipping.")
```

Comment:
Should the script fail if the model is not present? Dont we want all models ran?

Reply:
Yes, fixed